cranelift/codegen/src/isa/x64/inst/emit.rs

```diff
@@ -4693,6 +4692,19 @@ pub(crate) fn emit(
         Inst::DummyUse { .. } => {
             // Nothing.
         }
+
+        Inst::External { inst } => {
+            let mut known_offsets = [0, 0];
+            // These values are transcribed from is happening in
+            // `SyntheticAmode::finalize`. This, plus the `Into` logic converting a
+            // `SyntheticAmode` to its external counterpart, are
+            let frame = state.frame_layout();
+            known_offsets[external::offsets::KEY_INCOMING_ARG] =
+                i32::try_from(frame.tail_args_size + frame.setup_area_size).unwrap();
+            known_offsets[external::offsets::KEY_SLOT_OFFSET] =
+                i32::try_from(frame.outgoing_args_size).unwrap();
```

Comment:
@cfallin, I'm not a big fan of this `KnownOffsetTable` approach: it seems like we may want to just add appropriate `CodeSink`/`MachBuffer` methods to propagate this kind of thing.

Reply:
Perhaps! Were you thinking we'd have specific trait methods for e.g. `incoming_arg` and `slot_offset`? Or that we'd still be generic over the set of known offsets but have a query method?

I'm totally fine with the latter, in fact it's probably even cleaner; for the former I'd be concerned about baking details of Cranelift's ABI impl into the lower-level library and would still want to follow the analogy of a real assembler letting one define symbolic constants and use them with specific values plugged in. Happy to look at proposed diff if you have one.

Reply:
I agree, we should avoid Cranelift-izing this assembler as much as possible. I guess things I don't like about the known offset table approach is that (a) we rebuild this table for every emissions and (b) building it the way I'm doing it here feels quite fragile. How long before things get out of sync with the keys used during amode conversion?!